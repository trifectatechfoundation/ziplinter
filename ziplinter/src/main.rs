use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
