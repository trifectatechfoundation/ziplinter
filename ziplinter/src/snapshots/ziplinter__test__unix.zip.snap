---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 8,
        "crc32": 2098461837,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2176188416,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          168,
          139,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 438,
        "modified": "2011-12-08T10:04:24Z",
        "name": "hello",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 8
      },
      "local": {
        "accessed": null,
        "compressed_size": 8,
        "crc32": 2098461837,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          168,
          139,
          224,
          78,
          86,
          155,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:04:24Z",
        "name": "hello",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 1000,
        "uncompressed_size": 8
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 6,
        "crc32": 2055117726,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2176188416,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          194,
          139,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 71,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 438,
        "modified": "2011-12-08T10:04:50Z",
        "name": "dir/bar",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 6
      },
      "local": {
        "accessed": null,
        "compressed_size": 6,
        "crc32": 2055117726,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          194,
          139,
          224,
          78,
          48,
          154,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:04:50Z",
        "name": "dir/bar",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 1000,
        "uncompressed_size": 6
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 0,
        "crc32": 0,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 1107230736,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          134,
          140,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 142,
        "internal_attrs": 0,
        "method": "Store",
        "mode": 2147484159,
        "modified": "2011-12-08T10:08:06Z",
        "name": "dir/empty/",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 0
      },
      "local": {
        "accessed": null,
        "compressed_size": 0,
        "crc32": 0,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          134,
          140,
          224,
          78,
          134,
          140,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 2147483648,
        "modified": "2011-12-08T10:08:06Z",
        "name": "dir/empty/",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 1000,
        "uncompressed_size": 0
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 12,
        "crc32": 3127775578,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2166620161,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          16,
          140,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 210,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 292,
        "modified": "2011-12-08T10:06:08Z",
        "name": "readonly",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 12
      },
      "local": {
        "accessed": null,
        "compressed_size": 12,
        "crc32": 3127775578,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          16,
          140,
          224,
          78,
          201,
          155,
          224,
          78,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2011-12-08T10:06:08Z",
        "name": "readonly",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 1000,
        "uncompressed_size": 12
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "comment": [],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 4,
        "directory_offset": 288,
        "directory_records": 4,
        "directory_size": 310,
        "disk_nbr": 0
      },
      "offset": 598
    },
    "dir64": null,
    "global_offset": 0
  },
  "size": 620
}
